### PR TITLE
Improve mate pair heuristic (resolves #436)

### LINKED ISSUE #436: Mate pair designator heuristic fails if parent directory name contains R1 or R2
From http://jenkins.cgcloud.info/job/toil-pull-request-integration-tests/78/testReport/junit/src.toil.test.provisioners.cgcloud.provisionerTest/CGCloudProvisionerTest/testStaticCluster/:

```
ip-172-31-21-199 2016-09-02 04:12:14,899 MainThread WARNING toil.leader: 90fb8c31-9a23-41c9-a383-abf39615a8c3:	Traceback (most recent call last):
ip-172-31-21-199 2016-09-02 04:12:14,899 MainThread WARNING toil.leader: 90fb8c31-9a23-41c9-a383-abf39615a8c3:	  File "/usr/local/lib/python2.7/dist-packages/toil/worker.py", line 331, in main
ip-172-31-21-199 2016-09-02 04:12:14,899 MainThread WARNING toil.leader: 90fb8c31-9a23-41c9-a383-abf39615a8c3:	    job._runner(jobWrapper=jobWrapper, jobStore=jobStore, fileStore=fileStore)
ip-172-31-21-199 2016-09-02 04:12:14,899 MainThread WARNING toil.leader: 90fb8c31-9a23-41c9-a383-abf39615a8c3:	  File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 1085, in _runner
ip-172-31-21-199 2016-09-02 04:12:14,899 MainThread WARNING toil.leader: 90fb8c31-9a23-41c9-a383-abf39615a8c3:	    returnValues = self._run(jobWrapper, fileStore)
ip-172-31-21-199 2016-09-02 04:12:14,899 MainThread WARNING toil.leader: 90fb8c31-9a23-41c9-a383-abf39615a8c3:	  File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 1033, in _run
ip-172-31-21-199 2016-09-02 04:12:14,899 MainThread WARNING toil.leader: 90fb8c31-9a23-41c9-a383-abf39615a8c3:	    return self.run(fileStore)
ip-172-31-21-199 2016-09-02 04:12:14,899 MainThread WARNING toil.leader: 90fb8c31-9a23-41c9-a383-abf39615a8c3:	  File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 1178, in run
ip-172-31-21-199 2016-09-02 04:12:14,899 MainThread WARNING toil.leader: 90fb8c31-9a23-41c9-a383-abf39615a8c3:	    rValue = userFunction(*((self,) + tuple(self._args)), **self._kwargs)
ip-172-31-21-199 2016-09-02 04:12:14,899 MainThread WARNING toil.leader: 90fb8c31-9a23-41c9-a383-abf39615a8c3:	  File "/tmp/pip-build-EwMPRA/toil-scripts/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py", line 201, in process_sample_tar
ip-172-31-21-199 2016-09-02 04:12:14,900 MainThread WARNING toil.leader: 90fb8c31-9a23-41c9-a383-abf39615a8c3:	  File "/tmp/pip-build-EwMPRA/toil-scripts/toil_scripts/lib/__init__.py", line 44, in require
ip-172-31-21-199 2016-09-02 04:12:14,900 MainThread WARNING toil.leader: 90fb8c31-9a23-41c9-a383-abf39615a8c3:	UserError: 
ip-172-31-21-199 2016-09-02 04:12:14,900 MainThread WARNING toil.leader: 90fb8c31-9a23-41c9-a383-abf39615a8c3:	
ip-172-31-21-199 2016-09-02 04:12:14,900 MainThread WARNING toil.leader: 90fb8c31-9a23-41c9-a383-abf39615a8c3:	Check fastq naming, uneven number of pairs found: r1: ['/var/lib/toil/toil-d4a0309f-4911-495e-a6ca-2d599eea49f7/tmpI63wE8/244bcf83-c584-4098-9109-96665b1c4972/tR2afmv/R1.fq.gz'], r2: ['/var/lib/toil/toil-d4a0309f-4911-495e-a6ca-2d599eea49f7/tmpI63wE8/244bcf83-c584-4098-9109-96665b1c4972/tR2afmv/R1.fq.gz', '/var/lib/toil/toil-d4a0309f-4911-495e-a6ca-2d599eea49f7/tmpI63wE8/244bcf83-c584-4098-9109-96665b1c4972/tR2afmv/R2.fq.gz', '/var/lib/toil/toil-d4a0309f-4911-495e-a6ca-2d599eea49f7/tmpI63wE8/244bcf83-c584-4098-9109-96665b1c4972/tR2afmv/sample.tar']
```

Note how the directory name `tR2afmv` contains the string `R2` which throws off the heuristic.

### PR DESCRIPTION
**Pull request by jvivian**

### CODE REVIEW

**Comment by jvivian** [comment]
@hannes-ucsc — The regex fails on an interesting use case:  Files named `R1.fq.gz` or `R2.fq.gz` return `None` from `re.search` because there's no `-`, `.` or `_`.

**Comment by hannes-ucsc** [comment]
Try replacing [._-] with (?:^|[._-]) assuming you have already applied basename. If not, just add a / inside the brackets (before the dash, the dash must come last).

**Comment by jvivian** [comment]
Thanks @hannes-ucsc — I had a fix but I'll ask about yours at scrum.

**Comment by jvivian** [comment]
@hannes-ucsc — Ready for review.

File: src/toil_scripts/rnaseq_cgl/README.md [hunk 1]
```diff
@@ -90,6 +90,9 @@ Alternatively, many samples can be placed in a manifest file created by using th
 All samples and inputs must be submitted as URLs with support for the following schemas: 
 `http://`, `file://`, `s3://`, `ftp://`.
 
+Samples consisting of tarballs with fastq files inside _must_ follow the file name convention of ending in an 
+R1/R2 or \_1/\_2 followed by one of the 4 extensions: `{ .fastq.gz, .fastq, .fq.gz, .fq }`.
```
Comment (hannes-ucsc) [inline src/toil_scripts/rnaseq_cgl/README.md]:
```
followed by `.fastq.gz`, `.fastq`, `.fq.gz` or `.fq.`.
```
?

File: src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py [hunk 2]
```diff
@@ -201,12 +202,18 @@ def process_sample(job, config, input_tar=None, input_r1=None, input_r2=None, gz
...
     if config.paired:
-        r1 = sorted([x for x in fastqs if 'R1' in x])
-        r2 = sorted([x for x in fastqs if 'R2' in x])
-        if not r1 or not r2:
-            r1 = sorted([x for x in fastqs if '_1' in x])
-            r2 = sorted([x for x in fastqs if '_2' in x])
-        require(len(r1) == len(r2), 'Check fastq naming, uneven number of pairs found: r1: {}, r2: {}'.format(r1, r2))
+        r1, r2 = [], []
+        # Pattern convention: Look for R1/R2 or _1/_2 at the end of files named .fastq.gz, .fastq, .fq.gz, or .fq
+        pattern = '(?:^|[._-])R?([12])(\\.fastq(\\.gz)?|\\.fq(\\.gz)?)$'
```
Comment (hannes-ucsc) [inline src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py]: pattern = re.compile( `…

File: src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py [hunk 3]
```diff
@@ -201,12 +202,18 @@ def process_sample(job, config, input_tar=None, input_r1=None, input_r2=None, gz
...
+        for fastq in sorted(fastqs):
+            match = re.search(pattern, os.path.basename(fastq))
```
Comment (hannes-ucsc) [inline src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py]:
```
match = pattern.search( os.…
```

File: src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py [hunk 4]
```diff
@@ -201,12 +202,18 @@ def process_sample(job, config, input_tar=None, input_r1=None, input_r2=None, gz
...
+            if not match:
+                raise UserError('Fastq file fails to meet required convention. See documentation: {}'.format(fastq))
```
Comment (hannes-ucsc) [inline src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py]:
```
FASTQ file name fails to meet required convention for paired reads (see documentation).
```

File: src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py [hunk 5]
```diff
@@ -201,12 +202,18 @@ def process_sample(job, config, input_tar=None, input_r1=None, input_r2=None, gz
...
+            elif match.group(1) == '1':
+                r1.append(fastq)
+            elif match.group(1) == '2':
+                r2.append(fastq)
```
Comment (hannes-ucsc) [inline src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py]:
```
else:
    assert False
```

**Comment by hannes-ucsc** [comment]
Done with review.

**Comment by jvivian** [comment]
@hannes-ucsc — Addressed review comments.

**Comment by hannes-ucsc** [comment]
+1

File: src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py [hunk 6]
```diff
@@ -201,12 +202,21 @@ def process_sample(job, config, input_tar=None, input_r1=None, input_r2=None, gz
...
+            match = pattern.search(os.path.basename(fastq))
+            if not match:
+                raise UserError('FASTQ file name fails to meet required convention for paired reads '
+                                '(see documentation). ' + fastq)
+            elif match.group(1) == '1':
+                r1.append(fastq)
+            elif match.group(1) == '2':
+                r2.append(fastq)
+            else:
+                assert False
```
Comment (fnothaft) [inline src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py]: I think it is correct to assert here (vs. `require`), but I'd give a meaningful error message.

Comment (fnothaft) [inline src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py]: At the least, the error message should indicate what the filename that failed in the regex was.

**Comment by fnothaft** [comment]
LGTM other than 1 nit.

File: src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py (hunk 6 above)
Comment (hannes-ucsc) [inline src/toil_scripts/rnaseq_cgl/rnaseq_cgl_pipeline.py]: I'd rather print just the value of match.group(1). The file name matched already so this indicates problem with the regex.

The trade-off here is between ease of diagnostics if the problem occurs on one hand and distracting verbosity and tedium on the other. I think a balance is struck with

```
assert False, match.group(1)
```

**Pull request merged**